src/V3Gate.cpp

```diff
@@ -919,16 +919,8 @@ class GateDedupeHash final : public V3DupFinderUserSame {
 
     VL_DEBUG_FUNC;  // Declare debug()
 
-    bool sameHash(AstNode* node1p, AstNode* node2p) {
-        return node1p  //
-               && node2p  //
-               && !node1p->sameHash().isIllegal()  //
-               && !node2p->sameHash().isIllegal()  //
-               && m_dupFinder.sameNodes(node1p, node2p);
-    }
-
     bool same(AstNode* node1p, AstNode* node2p) {
-        return node1p == node2p || sameHash(node1p, node2p);
+        return node1p == node2p || (node1p && node1p->sameTree(node2p));
```

Comment:
sameHash was O(1) for memoized trees, while sameTree is O(n^2).  So I'm worried that performance is potentially bad.  Am I missing something?

Reply:
This is what I was eluding to in the original filing for this PR, so here is how I see it. Applying the comparison on this line to a a set of n trees pairwise is O(n^2) in the number of comparisons (number of pairs). `AstNode::sameTree` itself, is O(size_of_tree) in the worst case, which happens if the operands of `sameTree` are indeed identical copies, which means this line is O(n^2*size_of_tree), iff you are comparing identical copies of the same tree. In practice the indentity comparison over the pointers, and the short circuiting in `sameTree` menas that for comparing the same tree instance to itself, or trees of different types/shapes is a lot closer to O(1), so this compariosn is Omega(n^2) and O(n^2*size_of_tree), and in practice as we are mostly coparing the same instance to itself or different trees, the complexity should be closer to the lower bound. 

Also if you see where this `same` fucntion is used within `isSame`, it's only ever comparing `AstActive` nodes, which are very likely not to compare equals (and for the purposes of V3Gate, we probably only care about them either being identical instances, or having the same sensitivities anyway, so if this becomes a problem, I'm happy to improve on it which should also speed things up), and `AstNodeMath` for if conditions, which are hopefully small, and to be safe they should probably be only considered same when identical instances (otherwise if writing the condition between 2 ifs don't really merge).

Reply:
That makes sense, if you could please add a comment there, especially to protect against if someone mistakenly copies that somewhere more dangerous in the future, thanks.

Reply:
Done